### PR TITLE
Fix handling of helpers with 5 parameters on Windows

### PR DESCRIPTION
**Pull request by Alan-Jowett**
Changes:
1) Permit passing additional metadata about the extension/helper function.
2) Interpreter - Correctly invoke helper function by type (fix case where function accepts 5th parameter).
3) Interpreter - Honor "unwind-on-success" flag.
2) Jitter - Windows x64 - Spill 5th parameter to stack.
3) Jitter - Windows x64 - Allocate stack space for home registers.
4) Jitter - Honor "unwind-on-success" flag

Signed-off-by: Alan Jowett <alan.jowett@microsoft.com>

### CODE REVIEW

**Comment by Alan-Jowett** [comment]
Note:
Draft until I can complete testing on Windows x64.

**Comment by Alan-Jowett** [comment]
Spilling R5 to the stack has no negative effects on Windows x64, so to simplify, always spill R5 to the stack.

File: vm/ubpf_jit_x86_64.c [hunk 1]
```diff
@@ -407,7 +407,14 @@ translate(struct ubpf_vm *vm, struct jit_state *state, char **errmsg)
         case EBPF_OP_CALL:
             /* We reserve RCX for shifts */
             emit_mov(state, RCX_ALT, RCX);
+            // Windows x64 ABI spills 5th parameter to stack
+#if defined(_WIN32)
```
[the remark this answers is not in the document]
Comment (Alan-Jowett) [reply vm/ubpf_jit_x86_64.c:411]: For historical reasons _WIN32 means "new" Windows (vs _WIN16 for the previous Windows architecture) with additional preprocessor definitions for each architecture:

From Windows.h, there are the following architectures defined:
1) _68K_ - Old Macintosh
2) _MPPC_ - PowerPC
3) _X86_ - 32bit x86 Intel
4) _IA64_ - 64bit Itanium Intel
5) _AMD64_ - 64bit x64 Intel / AMD 64
6) _ARM_ - 32bit ARM
7) _ARM64_ - 64bit ARM

All of these have _Win32 defined.

[the remark this answers is not in the document]
Comment (Alan-Jowett) [reply vm/ubpf_jit_x86_64.c:411]: Also note:
For 32 bit x86, there are actually 3 different calling conventions in Windows, some of which use registers others store it on the stack. There are also different rules for caller vs callee cleaning up the stack. 

[Windows 32 bit calling conventions](https://docs.microsoft.com/en-us/cpp/cpp/argument-passing-and-naming-conventions?view=msvc-160)

**Comment by Alan-Jowett** [comment]
Rebased on to correct master branch.

**Comment by Alan-Jowett** [comment]
> Can you please rebase this on the latest master? The first 4 commits of this PR have already been merged, so only the last one is really new.
> 
> Also, can you make the commit message more descriptive and explain why this is change is being done? The reasoning given on the linked issue makes sense, but it is very useful and convenient to have all of that information in the commit message, without having to jump through links and reading PR messages

File: vm/ubpf_jit_x86_64.c (hunk 1 above)
[the remark this answers is not in the document]
Comment (iomartin) [reply vm/ubpf_jit_x86_64.c:410]: Thanks for the clarification

File: vm/ubpf_jit_x86_64.h [hunk 2]
```diff
@@ -319,11 +319,22 @@ emit_store_imm32(struct jit_state *state, enum operand_size size, int dst, int32
 static inline void
 emit_call(struct jit_state *state, void *target)
 {
+#if defined(_WIN32)
+    /* Windows x64 ABI requires home register space */
+    /* Allocate home register space - 4 registers */
```
Comment (iomartin) [inline vm/ubpf_jit_x86_64.h:324]: Why not put the emit_push() here (and the emit_pop at the end)?

Comment (Alan-Jowett) [reply vm/ubpf_jit_x86_64.h:327]: Good point.

Comment (Alan-Jowett) [reply vm/ubpf_jit_x86_64.h:327]: I was originally only going to do this if there are 5 parameters. Given that I changed it to always do it (to keep the ubpf_register function simpler), I guess it makes sense to do it here.

Note:
We will need to modify ubpf_register to support x86 JIT (assuming we ever do) as it needs to know the exact size of of parameters to adjust the stack correctly).

Comment (iomartin) [reply vm/ubpf_jit_x86_64.h:327]: I see. This is going to be a performance hit for every call, but I don't know if there is an easy way in ubpf to know how many parameters a function has.

**Review by iomartin — COMMENTED** [review]
Can you please squash the two commits? Other than that it looks fine to me.

File: vm/ubpf_jit_x86_64.h (hunk 2 above)
Comment (Alan-Jowett) [reply vm/ubpf_jit_x86_64.h:327]: We could modify:
https://github.com/iovisor/ubpf/blob/843394752c752e38f79d6d49fc704a73ee6490bd/vm/inc/ubpf.h#L67

To accept a parameter count, store that in ubpf_vm and pass parameter count to:
 https://github.com/iovisor/ubpf/blob/843394752c752e38f79d6d49fc704a73ee6490bd/vm/ubpf_jit_x86_64.h#L320

But until I get more data on how bad the perf hit is, I was thinking of keeping the code simpler. 

Note:
This only affects uBPF on Windows. While Linux does spill 7th+ parameter to the stack, BPF only defines 5 parameters for BPF calls :-)

**Review by iomartin — APPROVED** [review]

**Pull request merged**